src/drivers/imx/dummy-dma.c

```diff
@@ -45,54 +111,279 @@ static int dummy_dma_status(struct dma_chan_data *channel,
 			    struct dma_chan_status *status,
 			    uint8_t direction)
 {
+	struct dma_chan_pdata *ch = dma_chan_get_data(channel);
+
+	(void)direction; /* TODO should we check it? */
+
+	status->state = channel->status;
+	status->flags = 0; /* TODO What flags should be put here? */
+	status->r_pos = ch->r_pos;
+	status->w_pos = ch->w_pos;
+
+	status->timestamp = 0; /* TODO what would be the correct one? */
 	return 0;
 }
 
 /* set the DMA channel configuration, source/target address, buffer sizes */
 static int dummy_dma_set_config(struct dma_chan_data *channel,
 				struct dma_sg_config *config)
 {
-	return 0;
+	struct dma_chan_pdata *ch = dma_chan_get_data(channel);
+	uint32_t flags;
+	int ret = 0;
+
+	spin_lock_irq(channel->dma->lock, flags);
+
+	if (!config->elem_array.count) {
+		trace_dummydma_error("dummy-dmac: %d channel %d no DMA descriptors",
+				     channel->dma->plat_data.id,
+				     channel->index);
+
+		ret = -EINVAL;
+		goto out;
+	}
+
+	channel->direction = config->direction;
+	/* TODO Should we keep this check? */
+	if (config->direction != DMA_DIR_HMEM_TO_LMEM &&
+	    config->direction != DMA_DIR_LMEM_TO_HMEM) {
+		/* Shouldn't even happen though */
+		trace_dummydma_error("dummy-dmac: %d channel %d invalid direction %d",
+				     channel->dma->plat_data.id, channel->index,
+				     config->direction);
+		ret = -EINVAL;
+		goto out;
+	}
+	channel->desc_count = config->elem_array.count;
+	ch->elems = &config->elem_array;
+	ch->sg_elem_curr_idx = 0;
+
+	channel->status = COMP_STATE_PREPARE;
+out:
+	spin_unlock_irq(channel->dma->lock, flags);
+	return ret;
 }
 
 /* restore DMA context after leaving D3 */
 static int dummy_dma_pm_context_restore(struct dma *dma)
 {
+	/* Virtual device, no hardware registers */
 	return 0;
 }
 
 /* store DMA context after leaving D3 */
 static int dummy_dma_pm_context_store(struct dma *dma)
 {
+	/* Virtual device, no hardware registers */
 	return 0;
 }
 
 static int dummy_dma_set_cb(struct dma_chan_data *channel, int type,
 		void (*cb)(void *data, uint32_t type, struct dma_cb_data *next),
 		void *data)
 {
+	channel->cb = cb;
+	channel->cb_data = data;
+	channel->cb_type = type;
 	return 0;
 }
 
+static bool get_next_elem(struct dma_chan_data *channel,
+			  struct dma_cb_data *next)
+{
+	struct dma_chan_pdata *ch = dma_chan_get_data(channel);
+	int curr_index = ch->sg_elem_curr_idx++;
+
+	if (curr_index >= channel->desc_count)
+		return false;
+
+	next->elem = ch->elems->elems[curr_index];
+	return true;
+}
+
+static int docopy(struct dma_chan_data *ch, struct dma_cb_data *next)
+{
+	struct dma_chan_pdata *pdata = dma_chan_get_data(ch);
+	/* Perform a single elem copy */
+	void *dest = (void *)next->elem.dest;
+	void *src = (void *)next->elem.src;
```

Comment:
This only works correctly with HOST_PTABLE config enabled though, so should I add the local changes I made to enable HOST_PTABLE on this platform or figure something else out? With HOST_PTABLE config disabled, I'm getting NULL as the first SG element source.

Reply:
Should we make dummy-dma config depend on HOST_PTABLE?

Reply:
@dbaluta Just noticed this comment, updated.